validator/core/src/input/proto.rs

```diff
@@ -0,0 +1,205 @@
+        "uint64"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Unsigned(0)
+    }
+
+    fn proto_primitive_data(&self) -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Unsigned(*self)
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
+    }
+}
+
+impl traits::ProtoPrimitive for i32 {
+    fn proto_primitive_type() -> &'static str {
+        "int32"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Signed(0)
+    }
+
+    fn proto_primitive_data(&self) -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Signed((*self).into())
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
+    }
+}
+
+impl traits::ProtoPrimitive for i64 {
+    fn proto_primitive_type() -> &'static str {
+        "int64"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Signed(0)
+    }
+
+    fn proto_primitive_data(&self) -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Signed(*self)
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0
+    }
+}
+
+impl traits::ProtoPrimitive for f32 {
+    fn proto_primitive_type() -> &'static str {
+        "float"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Float(0.0)
+    }
+
+    fn proto_primitive_data(&self) -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Float((*self).into())
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0.0
+    }
+}
+
+impl traits::ProtoPrimitive for f64 {
+    fn proto_primitive_type() -> &'static str {
+        "double"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Float(0.0)
+    }
+
+    fn proto_primitive_data(&self) -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Float(*self)
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        *self != 0.0
+    }
+}
+
+impl traits::ProtoPrimitive for String {
+    fn proto_primitive_type() -> &'static str {
+        "string"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::String(String::new())
+    }
+
+    fn proto_primitive_data(&self) -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::String(self.clone())
+    }
+
+    fn proto_primitive_is_default(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl traits::ProtoPrimitive for Vec<u8> {
+    fn proto_primitive_type() -> &'static str {
+        "bytes"
+    }
+
+    fn proto_primitive_default() -> primitive_data::PrimitiveData {
+        primitive_data::PrimitiveData::Bytes(vec![])
```

Comment:
You could also use the Default trait for all these impls here.

Reply:
Sure, I guess, but Rust only "incidentally" has the same default value definitions that protobuf has, so I prefer explicitly stating what protobuf considers to be the default value for a type. Unless I'm misunderstanding what you mean?